File: app/src/main/java/com/itsvks/layouteditor/editor/initializer/AttributeMap.java
package com.itsvks.layouteditor.editor.initializer;

import java.util.ArrayList;

public class AttributeMap {
    private ArrayList<Attribute> attrs = new ArrayList<>();

    public void putValue(String key, String value) {
        if (contains(key)) {
            int index = getAttributeIndexFromKey(key);
            attrs.get(index).value = value;
        } else attrs.add(new Attribute(key, value));
    }

    public void removeValue(String key) {
        attrs.remove(getAttributeIndexFromKey(key));
    }

    public String getValue(String key) {
        Attribute attr = attrs.get(getAttributeIndexFromKey(key));
        return attr.value;
    }

    public ArrayList<String> keySet() {
        ArrayList<String> keys = new ArrayList<>();

        for (Attribute attr : attrs) {
            keys.add(attr.key);
        }

        return keys;
    }

    public ArrayList<String> values() {
        ArrayList<String> values = new ArrayList<>();

        for (Attribute attr : attrs) {
            values.add(attr.value);
        }

        return values;
    }

    public boolean contains(String key) {
        for (Attribute attr : attrs) {
            if (attr.key.equals(key)) return true;
        }

        return false;
    }

    private int getAttributeIndexFromKey(String key) {
        int index = 0;

        for (Attribute attr : attrs) {
            if (attr.key.equals(key)) return index;
            index++;
        }

        return index;
    }

    private class Attribute {
        private String key, value;

        public Attribute(String key, String value) {
            this.key = key;
            this.value = value;
        }
    }
}

File: app/src/main/java/com/itsvks/layouteditor/editor/initializer/AttributeMap.kt
package com.itsvks.layouteditor.editor.initializer

class AttributeMap {
  private val attrs: MutableList<Attribute> = ArrayList()

  /**
   * Puts a key-value pair into the AttributeMap
   *
   * @param key the key of the attribute
   * @param value the value of the attribute
   */
  fun putValue(key: String, value: String) {
    if (contains(key)) {
      val index = getAttributeIndexFromKey(key)
      attrs[index].value = value
    } else {
      attrs.add(Attribute(key, value))
    }
  }

  /**
   * Removes a key-value pair from the AttributeMap
   *
   * @param key the key of the attribute to be removed
   */
  fun removeValue(key: String) {
    val index = getAttributeIndexFromKey(key)
    attrs.removeAt(index)
  }

  /**
   * Gets the value associated with the given key in the AttributeMap
   *
   * @param key the key of the attribute
   * @return the value of the attribute
   */
  fun getValue(key: String): String {
    val index = getAttributeIndexFromKey(key)
    val attr = attrs[index]
    return attr.value
  }

  /**
   * Gets a list of all the keys in the AttributeMap
   *
   * @return a list of all keys
   */
  fun keySet(): List<String> {
    val keys: MutableList<String> = ArrayList()

    for (attr in attrs) {
      keys.add(attr.key)
    }

    return keys
  }

  /**
   * Gets a list of all the values in the AttributeMap
   *
   * @return a list of all values
   */
  fun values(): List<String> {
    val values: MutableList<String> = ArrayList()

    for (attr in attrs) {
      values.add(attr.value)
    }

    return values
  }

  /**
   * Checks if the AttributeMap contains a key
   *
   * @param key the key to check for
   * @return true if the AttributeMap contains the key, false otherwise
   */
  fun contains(key: String): Boolean {
    for (attr in attrs) {
      if (attr.key == key) {
        return true
      }
    }

    return false
  }

  /**
   * Gets the index of the Attribute with the given key
   *
   * @param key the key of the Attribute
   * @return the index of the Attribute
   */
  private fun getAttributeIndexFromKey(key: String): Int {
    var index = 0

    for (attr in attrs) {
      if (attr.key == key) {
        return index
      }

      index++
    }

    return index
  }

  private class Attribute
  /**
   * Constructs an Attribute with the specified key-value pair
   *
   * @param key the key of the Attribute
   * @param value the value of the Attribute
   */(val key: String, var value: String)
}
